File: app/src/main/java/com/example/mike/mp3player/commons/MediaItemUtils.java
package com.example.mike.mp3player.commons;

import android.net.Uri;
import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.io.File;

import static android.support.v4.media.MediaBrowserCompat.MediaItem;
import static android.support.v4.media.MediaMetadataCompat.METADATA_KEY_ALBUM_ART_URI;
import static android.support.v4.media.MediaMetadataCompat.METADATA_KEY_ARTIST;
import static android.support.v4.media.MediaMetadataCompat.METADATA_KEY_DURATION;
import static com.example.mike.mp3player.commons.Constants.LIBRARY_ID;
import static com.example.mike.mp3player.commons.Constants.MEDIA_ITEM_TYPE;
import static com.example.mike.mp3player.commons.Constants.ROOT_ITEM_TYPE;
import static com.example.mike.mp3player.commons.MetaDataKeys.META_DATA_DIRECTORY;

public final class MediaItemUtils {

    public static boolean hasExtras(MediaItem item) {
        return item != null && item.getDescription().getExtras() != null;
    }


    public static boolean hasTitle(MediaItem item) {
        return item != null && item.getDescription().getTitle() != null;
    }

    public static boolean hasDescription(MediaItem item) {
        return item != null && item.getDescription().getDescription() != null;
    }

    public static Bundle getExtras(MediaItem item) {
        if (!hasExtras(item)) {
            return null;
        }
        return item.getDescription().getExtras();
    }

    public static Object getExtra(String key, MediaItem item) {
        if (item == null ) {
            return null;
        }
        final Bundle extras = item.getDescription().getExtras();
        return null == extras ? null : extras.get(key);
    }

    public static String getMediaId(MediaItem item) {
        if (null != item) {
            return item.getDescription().getMediaId();
        }
        return null;
    }

    public static String getTitle(MediaItem i) {
        if (hasTitle(i)) {
            return i.getDescription().getTitle().toString();
        }
        return null;
    }

    public static String getDescription(MediaItem item) {
        if (hasDescription(item)) {
            return item.getDescription().getDescription().toString();
        }
        return null;
    }

    public static boolean hasExtra(String key, MediaItem item) {
        return hasExtras(item) && item.getDescription().getExtras().containsKey(key);
    }

    public static String getArtist(MediaItem item) {
         return (String) getExtra(METADATA_KEY_ARTIST, item);
    }

    public static String getAlbumArtPath(MediaItem item) {
        if (hasExtra(METADATA_KEY_ALBUM_ART_URI, item)) {
            Uri uri = (Uri) getExtra(METADATA_KEY_ALBUM_ART_URI, item);
            if (uri != null) {
                return uri.toString();
            }
        }
        return null;
    }

    public static String getDirectoryName(MediaItem item) {
        if (hasExtra(META_DATA_DIRECTORY, item)) {
            File directory = (File) getExtra(META_DATA_DIRECTORY, item);
            if (null != directory) {
                return directory.getName();
            }
        }
        return null;
    }

    public static String getDirectoryPath(MediaItem item) {
        if (hasExtra(META_DATA_DIRECTORY, item)) {
            File directory = (File) getExtra(META_DATA_DIRECTORY, item);
            if (null != directory) {
                return directory.getAbsolutePath();
            }
        }
        return null;
    }

    public static Uri getMediaUri(MediaItem item) {
        return item.getDescription().getMediaUri();
    }

    public static long getDuration(MediaItem item) {
        return (long) getExtra(METADATA_KEY_DURATION, item);
    }

    public static MediaItemType getMediaItemType(MediaItem item) {
        return (MediaItemType)getExtra(MEDIA_ITEM_TYPE, item);
    }

    public static String getLibraryId(MediaItem item) {
        return (String) getExtra(LIBRARY_ID, item);
    }

    public static MediaItemType getRootMediaItemType(MediaItem item) {
        return (MediaItemType)getExtra(ROOT_ITEM_TYPE, item);
    }

    public static Uri getAlbumArtUri(@NonNull MediaItem song) {
        Bundle extras = song.getDescription().getExtras();
        if (null != extras) {
            return  (Uri) extras.get(METADATA_KEY_ALBUM_ART_URI);
        }
        return null;
    }

    @Nullable
    public static String getRootTitle(@NonNull MediaItem song) {
        Bundle extras = song.getDescription().getExtras();
        if (null != extras) {
            MediaItemType mediaItemType = (MediaItemType) extras.getSerializable(ROOT_ITEM_TYPE);
            return mediaItemType == null ? null : mediaItemType.getTitle();
        }
        return null;
    }
}

File: commons/src/main/java/com/github/goldy1992/mp3player/commons/MediaItemUtils.kt
package com.github.goldy1992.mp3player.commons

import android.annotation.TargetApi
import android.net.Uri
import android.os.Build
import android.os.Build.VERSION_CODES.TIRAMISU
import android.os.Bundle
import android.util.Log
import androidx.media3.common.MediaItem
import com.github.goldy1992.mp3player.commons.Constants.EMPTY_MEDIA_ITEM_ID
import com.github.goldy1992.mp3player.commons.VersionUtils.isTiramisuOrHigher
import java.io.File
import java.io.Serializable

object MediaItemUtils : LogTagger {
    private fun hasExtras(item: MediaItem?): Boolean {
        return item != null && item.mediaMetadata.extras != null
    }

    private fun hasTitle(item: MediaItem?): Boolean {
        return item != null && item.mediaMetadata.title != null
    }

    private fun hasDescription(item: MediaItem?): Boolean {
        return item != null && item.mediaMetadata.description != null
    }

    @JvmStatic
    fun getExtras(item: MediaItem): Bundle? {
        return if (!hasExtras(item)) {
            null
        } else item.mediaMetadata.extras
    }

    @Suppress("DEPRECATION")
    fun getExtra(key: String?, item: MediaItem?): Any? {
        if (item == null) {
            return null
        }
        val extras = item.mediaMetadata.extras
        return extras?.get(key)
    }


    @Suppress("DEPRECATION")
    private fun <T : Serializable> getSerializableExtra(key: String?, clazz: Class<T>, item: MediaItem?): Any? {
        if (item == null) {
            return null
        }
        val extras = item.mediaMetadata.extras
        return if (isTiramisuOrHigher()) {
            extras?.getSerializable(key, clazz)
        } else {
            extras?.getSerializable(key)
        }
    }

    fun getIntExtra(key: String?, item: MediaItem?): Int? {
        if (item == null) {
            return null
        }
        val extras = item.mediaMetadata.extras
        return extras?.getInt(key)
    }

    fun getStringExtra(key: String?, item: MediaItem?): String? {
        if (item == null) {
            return null
        }
        val extras = item.mediaMetadata.extras
        return extras?.getString(key)
    }

    @JvmStatic
    fun getMediaId(item: MediaItem): String {
        return item.mediaId
    }

    @JvmStatic
    fun getTitle(mediaItem: MediaItem): String {
        return if (hasTitle(mediaItem)) {
            mediaItem.mediaMetadata.title.toString()
        } else Constants.UNKNOWN
    }

    @JvmStatic
    fun getDescription(item: MediaItem): String? {
        return if (hasDescription(item)) {
            item.mediaMetadata.description.toString()
        } else null
    }

    private fun hasExtra(key: String?, item: MediaItem?): Boolean {
        return hasExtras(item) && item?.mediaMetadata?.extras!!.containsKey(key)
    }

    private fun hasArtist(mediaItem : MediaItem) : Boolean {
        return mediaItem.mediaMetadata.artist != null
    }

    private fun hasFileCount(mediaItem : MediaItem?) : Boolean {
        return hasExtras(mediaItem) && hasExtra(Constants.FILE_COUNT, mediaItem)
    }

    fun getArtist(item: MediaItem): String {
        return if (hasArtist(item)) {
            item.mediaMetadata.artist.toString()
        } else {
            Constants.UNKNOWN
        }
    }

    @JvmStatic
    fun getAlbumArtPath(item: MediaItem): String? {
        return item.mediaMetadata.artworkUri?.toString()
    }

    @JvmStatic
    fun getDirectoryName(item: MediaItem?): String {
        if (hasExtra(MetaDataKeys.META_DATA_DIRECTORY, item)) {
            val directory = getSerializableExtra(MetaDataKeys.META_DATA_DIRECTORY, File::class.java, item) as File?
            if (null != directory) {
                return directory.name
            }
        }
        return Constants.UNKNOWN
    }

    @JvmStatic
    fun getDirectoryPath(item: MediaItem?): String {
        if (hasExtra(MetaDataKeys.META_DATA_DIRECTORY, item)) {
            val directory = getSerializableExtra(MetaDataKeys.META_DATA_DIRECTORY, File::class.java, item) as File?
            if (null != directory) {
                return directory.absolutePath
            }
        }
        return Constants.UNKNOWN
    }

    @JvmStatic
    fun getDirectoryUri(item: MediaItem?): Uri? {
        if (hasExtra(MetaDataKeys.META_DATA_DIRECTORY, item)) {
            val directory = getSerializableExtra(MetaDataKeys.META_DATA_DIRECTORY, File::class.java, item) as File?
            if (null != directory) {
                return Uri.fromFile(directory)
            }
        }
        return null
    }

    @JvmStatic
    fun getMediaUri(item: MediaItem): Uri? {
        return item.localConfiguration?.uri
    }

    @JvmStatic
    fun getDuration(item: MediaItem): Long {
        Log.i(logTag(), "item: $item")
        Log.i(logTag(), "metadata: ${item.mediaMetadata}")
        Log.i(logTag(), "extras: ${item.mediaMetadata.extras}")
        Log.i(logTag(), "duration: ${item.mediaMetadata.extras?.getLong(MetaDataKeys.DURATION)}")
        return item.mediaMetadata.extras?.getLong(MetaDataKeys.DURATION) ?: 0L
    }

    @JvmStatic
    fun getFileCount(item : MediaItem?) : Int {
        return if (hasFileCount(item)) getIntExtra(Constants.FILE_COUNT, item) ?: -1 else -1
    }

    @Suppress("DEPRECATION")
    @JvmStatic
    fun getMediaItemType(item: MediaItem): MediaItemType {
        val mediaItemType : MediaItemType? = if (Build.VERSION.SDK_INT >= TIRAMISU) {
            item.mediaMetadata.extras?.getSerializable(
                Constants.MEDIA_ITEM_TYPE,
                MediaItemType::class.java
            )
        } else {
            item.mediaMetadata.extras?.getSerializable(
                Constants.MEDIA_ITEM_TYPE) as MediaItemType?
        }
        if (mediaItemType == null) {
            Log.w(logTag(), "no MediaItemType found for item ${item.mediaId}")
        }
        return mediaItemType ?: MediaItemType.NONE
    }

    @JvmStatic
    fun getRootMediaItemType(item: MediaItem?): MediaItemType? {
        return getSerializableExtra(Constants.ROOT_ITEM_TYPE, MediaItemType::class.java, item) as MediaItemType?
    }

    fun getAlbumArtUri(song: MediaItem): Uri? {
        return song.mediaMetadata.artworkUri
    }

    @JvmStatic
    fun getAlbumArtImage(song: MediaItem): ByteArray? {
        return song.mediaMetadata.artworkData
    }

    @JvmStatic
    fun getEmptyMediaItem() : MediaItem {
        return MediaItemBuilder(EMPTY_MEDIA_ITEM_ID).build()
    }

    fun isEmptyMediaItem(mediaItem: MediaItem?) : Boolean {
        return mediaItem?.mediaId == EMPTY_MEDIA_ITEM_ID
    }

    fun noResultsFound(mediaItems : List<MediaItem>) : Boolean {
        return mediaItems.size == 1 && isEmptyMediaItem(mediaItems[0])
    }

    fun getAlbumTitle(mediaItem: MediaItem) : String {
        return mediaItem.mediaMetadata.albumTitle.toString()
    }

    fun getAlbumArtist(mediaItem: MediaItem) : String {
        return mediaItem.mediaMetadata.albumArtist.toString()
    }

    fun getAlbumRecordingYear(mediaItem: MediaItem) : String {
        return mediaItem.mediaMetadata.recordingYear.toString()
    }

    fun getAlbumReleaseYear(mediaItem: MediaItem) : String {
        return mediaItem.mediaMetadata.releaseYear.toString()
    }


    override fun logTag(): String {
        return "MediaItemUtils"
    }
}